### ISSUE TITLE
RustBinds does not contain expected methods

### ISSUE THREAD

**Issue by craigfay** [iduartgomez/rustypy#7]
Documentation examples to create Rust -> Python bindings don't behave as expected.
The `RustBinds` object is created, but does not contain methods from the Rust target library.

## Python Codebase

```python
# main.py

from rustypy.rswrapper import bind_rs_crate_funcs
    
source_path = "/rust_code/"
compiled_lib = "/rust_code/target/debug/libsome_lib.so"
lib_binds = bind_rs_crate_funcs(source_path, compiled_lib, prefixes=[])

# AttributeError: 'RustBinds' object has no attribute 'python_bind_example'
print(lib_binds.python_bind_example())
```

## Rust Codebase

```rust
// /rust_code/src/lib.rs

#[no_mangle]
pub extern "C" fn python_bind_example() -> Vec<i32> {
    vec![1,2,3]
}
```

```toml
# /rust_code/Cargo.toml
[package]
name = "rust_code"
version = "0.1.0"
authors = ["user"]
edition = "2018"

[lib]
name = "some_lib"
crate-type = ["cdylib"]
```

## Tests
I've cloned `rustypy`, and run `pytest` from the root directory with the following output:

```
============================================================================================================ test session starts =============================================================================================================
platform darwin -- Python 3.7.6, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /Users/craigfay/Repositories/rustypy
collected 0 items / 2 errors                                                                                                                                                                                                                 

=================================================================================================================== ERRORS ===================================================================================================================
__________________________________________________________________________________________________ ERROR collecting tests/test_py_to_rs.py ___________________________________________________________________________________________________
ImportError while importing test module '/Users/craigfay/Repositories/rustypy/tests/test_py_to_rs.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/test_py_to_rs.py:7: in <module>
    from rustypy.pywrapper import RustFuncGen
E   ModuleNotFoundError: No module named 'rustypy'
__________________________________________________________________________________________________ ERROR collecting tests/test_rs_to_py.py ___________________________________________________________________________________________________
ImportError while importing test module '/Users/craigfay/Repositories/rustypy/tests/test_rs_to_py.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/test_rs_to_py.py:7: in <module>
    from rustypy.rswrapper import Float, Double, Tuple
E   ModuleNotFoundError: No module named 'rustypy'
========================================================================================================== short test summary info ===========================================================================================================
ERROR tests/test_py_to_rs.py
ERROR tests/test_rs_to_py.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================================================= 2 errors in 0.19s ==============================================================================================================
FVFYK1B6HV2H:rustypy craigfay$ 

```

## Replication Steps
* Debian GNU/Linux 10 (buster)
* Python 3.7.3
* rustc 1.41.0 (5e1a79984 2020-01-27)
* cargo 1.41.0 (626f0f40e 2019-12-03)

You can reproduce the error in my my exact environment  by cloning [rustypy_debug](https://github.com/craigfay/rustypy_debug) and following instructions in its `README.md`. 

If the documentation is misleading or this is a genuine bug, I'm happy to dig in and help with a PR.

**Comment by iduartgomez** [comment]
Thanks for the report, hopefully will take a look this week.

**Comment by iduartgomez** [comment]
I have reproduced the issue both in the container and in my local environment with a fresh virtualenv. 

Running the tests on a fresh virtualenv worked though (maybe if you perform `pip install .` from the source it would work with pytest, this is what is done in CI and local dev). Nevertheless it doesn't matter as I could reproduce it in local dev, will check what is going on.

**Comment by iduartgomez** [comment]
I've identified the issue. For this particular case of the provided example library (a single extern C fn), the `syn` parser does not seem to like the content of the library , what ends up happening is that internally syn catches an unwinding panic and then returns, but without any parsed content whatsoever!

This means the crate data is never visited by the parser and hence no fn calls are generated... but it returns successfully. Gonna experiment a bit more to be sure this is the only problem. The fix in this case IMO would be to ensure that there have been any bindings generated (the caller is expecting at least one binding), and if there are none, then propagate the error. Additionally maybe document this behavior so is not unexpected in the future.

This would be less cryptic than just an `AttributeError`.

EDIT: The unwinding happens always and I got confused cause I was not passing the prefix appropriately, it actually parses the function but I am not generating the binds correctly as the type is not being appropriately annotated.

**Comment by iduartgomez** [comment]
@craigfay I failed to see an other problem, you cannot return a `Vec<i32>` here, you must return a raw pointer with the wrapper type, *mut PyList (so is required to use the rustypy library in the Rust side too) in Rust, and then in Python bind it appropriately specifying the return type.

Check `tests.test_rs_to_py.GenerateRustToPythonBinds.test_list_conversion` test for an example, maybe this is not sufficiently explained in the Wiki, neither in the "Type conversions#Calling Rust from Python" section nor the "Calling Rust from Python" one and should point to those examples or write a standalone full example. Also how to link `rustypy` in Rust is not pointed either...

PR to the Wiki to improve all this is welcome.

Ideally I wanted to write a proc macro to handle all the FFI automatically and generate the extern function calls automatically, with a lot of work this could also be handled automatically from the Python side, but for now it must be done manually by the caller.

**Comment by iduartgomez** [comment]
Discovered the error on the python side, you are passing an empty list for prefixes which shouldn't be allowed, have to catch this and raise an error.

**Comment by iduartgomez** [comment]
@craigfay I uploaded a new version with a couple fixes and also made a PR (https://github.com/craigfay/rustypy_debug/pull/1/) to your example repo with a couple changes needed to fix it. If you rebuild the docker image from scratch it should work now.

A PR to improve documentation/examples for clarity here is also welcome.

I uploaded the bridging library to crates.io for easier linking, the whole process now is a bit inefficient though as it's compiled twice (once in Python and other in Rust), so could be streamlined in the future probably.
